### DISCUSSION TITLE
How to use search functionality with custom theme?

### DISCUSSION THREAD

**Discussion by ovflowd** [shuding/nextra#1684] [Q&A]
Hey there 👋 I'd like to ask how can we adopt the built-in search functionality when using custom themes. Ty!

**Comment by jonny-snippyly** [comment]
Would also like to know. At first glance, seems like we could possibly re-use the search-related code from `nextra-theme-docs`, but you would have to figure out how things are wired up first, and then import them into your theme/layout. I don't think that package is setup to have things modularly picked and used. Might be more worth it to just implement Flexsearch with an input + autocomplete of your choice

**Comment by cglacet** [comment]
I would also be interesed by using the integrated search, but only to have it display in the body of a given page.

**Comment by alexrabin** [comment]
Here's how I got it working so far:

1. Enable flexsearch in next.config.js
```javascript

const withNextra = require("nextra")({
  theme: "./CustomNextraTheme.tsx",
  flexsearch: true,
});

```

2.  Install flexsearch `npm i flexsearch` if you are using typescript run `npm i @types/flexsearch -D` and make sure to run this command because the flexsearch types are out of date: `rm -rf node_modules/flexsearch/index.d.ts`

3. Copy over these files into your project and add your styles accordingly: https://github.com/shuding/nextra/blob/main/packages/nextra-theme-docs/src/components/flexsearch.tsx  https://github.com/shuding/nextra/blob/main/packages/nextra-theme-docs/src/components/search.tsx 
https://github.com/shuding/nextra/blob/main/packages/nextra-theme-docs/src/components/highlight-matches.tsx

4. Add the \<FlexSearch /> component into your NavigationBar component
5. You should be good to go

**Reply by officialrajdeepsingh** [reply]
Thank you for saving my life; I know nextra use `flexsearch`, but I can't find how to enable it.

**Reply by alexrabin** [reply]
@officialrajdeepsingh You're still struggling to set it up? Did the steps above not work for you?

**Reply by officialrajdeepsingh** [reply]
Yes @alexrabin it is too complicated to reuse the old `search.tsx`, `highlight-matches.tsx`, and `flexsearch.tsx` code. for example, if someone copies `search.tsx` it depends on another file. To solve those issues you need to copy those files also from the nextra docs theme. In this way, you copy a lot of code from the Nextra docs theme.
Now day flexsearch is not a single option I use [Orama](https://docs.oramasearch.com/) and my second option is to use a ready-made [Kbar](https://kbar.vercel.app/) component for search.

**Reply by alexrabin** [reply]
@officialrajdeepsingh  I was able to copy this files just fine and pick and choose what I needed. I'll write an article on how to implement flex search with a custom nextjs theme one of these days.

**Reply by ansi2u** [reply]
@alexrabin : if I'm using Nextra docs theme, do I have to still copy these files?

**Reply by alexrabin** [reply]
@ansi2u If you are using the nextra-docs-theme then this thread shouldn't have to apply to you

**Reply by devjiwonchoi** [reply]
@alexrabin Could you share your article?
I'm struggling with indexing the data since I'm importing the contents from another mdx, for example:

```mdx
// pages/index.mdx
import Content from 'docs/index.mdx'

<Content />
```
and it seems like this is not indexed to be caught on current flexsearch configuration..

**Reply by alexrabin** [reply]
@devjiwonchoi I haven't had time to create the article and I don't know when I'll be able to prioritize it 😕

**Comment by dylanfox5** [comment]
Experiencing issues with @alexrabin 's approach...having to copy a lot of code. Perhaps i'm going about it the wrong way?

I'm using the nextra docs theme, but I've integrated Nextra with an existing NextJS app. Therefore, I have an existing Navbar that I'd like to keep, but I'd like to integrate the search functionality from Nextra into it. Ideally I could simply import a search component and drop it in my Navbar. Any thoughts?
